### PR TITLE
chore[python]: Turn on mypy strict mode

### PR DESCRIPTION
**Pull request by zundertj**
See the comment in pyproject.toml. It basically means we have not changed that much, as we still ignore the warning. But there is not really a feasible way until PyO3 supports generating type annotations. Otherwise, maintaining stubs would just be a huge waste of resources, as it is duplicating the "outer" python api.

Ticks the last box on the mypy list in #4044

### CODE REVIEW

**Comment by stinodego** [comment]
No need to add this to overrides, we can just do:

```toml
[tool.mypy]
strict = true
warn_return_any = false
```

I agree that manually adding stub files is overkill.

**Comment by zundertj** [comment]
That would override it for all modules, i.e. also numpy, pandas, etc. I would prefer to make the selection as small as possible.

**Comment by stinodego** [comment]
> That would override it for all modules, i.e. also numpy, pandas, etc. I would prefer to make the selection as small as possible.

You're right; didn't think this through 😄 

We still need the ignore imports override though. So you could add an additional section of overrides with just the Polars package and the `warn_return_any` in it.

File: py-polars/pyproject.toml [hunk 1]
```diff
@@ -53,6 +39,15 @@ enable_error_code = [
 module = ["pyarrow.*", "polars.polars", "matplotlib.*", "fsspec.*", "connectorx", "IPython.*"]
 ignore_missing_imports = true
 
+[[tool.mypy.overrides]]
+module = ["polars.*", "fsspec.*", "pyarrow.*"]
```
Comment (stinodego) [inline py-polars/pyproject.toml:43]: I believe fsspec and pyarrow are not needed here.

Also, how exactly is adding `polars.*` here different from just including it in the general settings? I believe it's exactly the same (except for the fact that the override doesn't count for the `tests` folder).

Comment (zundertj) [reply py-polars/pyproject.toml:43]: > I believe fsspec and pyarrow are not needed here.


Correct, after adding back the original missing imports, that is not needed anymore, so I have removed those.

> Also, how exactly is adding `polars.*` here different from just including it in the general settings? I believe it's exactly the same (except for the fact that the override doesn't count for the `tests` folder).

True, still more strict than turning off the warning altogether as it covers `tests` (although return-any is not very likely to encounter in a test suite I would think).

Comment (stinodego) [reply py-polars/pyproject.toml:43]: Fair enough, we can keep it as an override!

**Comment by zundertj** [comment]
Python 3.7 test is fixed in #4416.

**Comment by stinodego** [comment]
I think you forgot to push your latest changes removing `fsspec` and `pyarrow` from the overrides. The Python test fix has been merged, so with a rebase + push I think we're good to go!

**Review by stinodego — APPROVED** [review]
Awesome! 🚀

**Comment by ritchie46** [comment]
That's quite an achievement. Well done. :+1:

**Pull request merged**